core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroxUpgradeService.java

```diff
@@ -0,0 +1,126 @@
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.tags.upgrade.UpgradeV4Request;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.image.ImageInfoV4Response;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.upgrade.UpgradeV4Response;
+import com.sequenceiq.cloudbreak.auth.CMLicenseParser;
+import com.sequenceiq.cloudbreak.auth.JsonCMLicense;
+import com.sequenceiq.cloudbreak.auth.PaywallAccessChecker;
+import com.sequenceiq.cloudbreak.auth.altus.Crn;
+import com.sequenceiq.cloudbreak.auth.altus.EntitlementService;
+import com.sequenceiq.cloudbreak.auth.altus.GrpcUmsClient;
+import com.sequenceiq.cloudbreak.core.flow2.service.ReactorFlowManager;
+import com.sequenceiq.cloudbreak.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.logger.MDCUtils;
+import com.sequenceiq.cloudbreak.service.StackCommonService;
+import com.sequenceiq.cloudbreak.service.image.ImageChangeDto;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.flow.api.model.FlowIdentifier;
+
+@Service
+public class DistroxUpgradeService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DistroxUpgradeService.class);
+
+    @Value("${cb.paywall.url}")
+    private String paywallUrl;
+
+    @Inject
+    private DistroxUpgradeAvailabilityService upgradeAvailabilityService;
+
+    @Inject
+    private EntitlementService entitlementService;
+
+    @Inject
+    private GrpcUmsClient umsClient;
+
+    @Inject
+    private PaywallAccessChecker paywallAccessChecker;
+
+    @Inject
+    private DistroxUpgradeImageSelector imageSelector;
+
+    @Inject
+    private StackCommonService stackCommonService;
+
+    @Inject
+    private ReactorFlowManager reactorFlowManager;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private CMLicenseParser cmLicenseParser;
+
+    public UpgradeV4Response triggerUpgrade(NameOrCrn cluster, Long workspaceId, String userCrn, UpgradeV4Request request) {
+        UpgradeV4Response upgradeV4Response = upgradeAvailabilityService.checkForUpgrade(cluster, workspaceId, request, userCrn);
+        validateUpgradeCandidates(cluster, upgradeV4Response);
+        verifyPaywallAccess(userCrn, request);
+        return initUpgrade(request, upgradeV4Response.getUpgradeCandidates(), cluster, workspaceId);
+    }
+
+    private UpgradeV4Response initUpgrade(UpgradeV4Request request, List<ImageInfoV4Response> upgradeCandidates, NameOrCrn cluster, Long workspaceId) {
+        ImageInfoV4Response image = imageSelector.determineImageId(request, upgradeCandidates);
+        ImageChangeDto imageChangeDto = createImageChangeDto(cluster, workspaceId, image);
+        Long stackId = stackService.getIdByNameOrCrnInWorkspace(cluster, workspaceId);
+        FlowIdentifier flowIdentifier = reactorFlowManager.triggerDistroxUpgrade(stackId, imageChangeDto, false);
+        // TODO: reason nice solution
```

Comment:
What does this todo mean?

Reply:
in sdx the message is read from property file for some reason, but we can leave it this way or I can use the same methodology for reading the message